cli/azd/pkg/ext/hooks.go

```diff
@@ -0,0 +1,379 @@
+	if scriptConfig.Type == "" {
+		scriptType, err := inferScriptTypeFromFilePath(scriptConfig.Path)
+		if err != nil {
+			return nil, err
+		}
+
+		scriptConfig.Type = scriptType
+	}
+
+	switch scriptConfig.Type {
+	case ScriptTypeBash:
+		return bash.NewBashScript(h.commandRunner, h.cwd, h.envVars), nil
+	case ScriptTypePowershell:
+		return powershell.NewPowershellScript(h.commandRunner, h.cwd, h.envVars), nil
+	default:
+		return nil, fmt.Errorf(
+			"script type '%s' is not a valid option. Only Bash and powershell scripts are supported",
+			scriptConfig.Type,
+		)
+	}
+}
+
+func (h *CommandHooks) getScriptsForHook(prefix HookType, commands []string) []*ScriptConfig {
+	validHookNames := []string{}
+
+	for _, commandName := range commands {
+		// Convert things like `azd config list` => 'configlist`
+		commandName = strings.TrimPrefix(commandName, "azd")
+		commandName = strings.TrimSpace(commandName)
+		commandName = strings.ReplaceAll(commandName, " ", "")
+
+		validHookNames = append(validHookNames, strings.ToLower(string(prefix)+commandName))
+	}
+
+	allHooks := []*ScriptConfig{}
+	explicitHooks := h.getExplicitHooks(prefix, validHookNames)
+	implicitHooks := h.getImplicitHooks(prefix, validHookNames)
+
+	allHooks = append(allHooks, explicitHooks...)
+
+	// Only append implicit hooks that were not already wired up explicitly.
+	for _, implicitHook := range implicitHooks {
+		index := slices.IndexFunc(allHooks, func(hook *ScriptConfig) bool {
+			return implicitHook.Name == hook.Name
+		})
+
+		if index >= 0 {
+			log.Printf(
+				"Skipping command hook @ '%s'. An explicit hook for '%s' was already defined in azure.yaml.\n",
+				implicitHook.Path,
+				implicitHook.Name,
+			)
+			continue
+		}
+
+		allHooks = append(allHooks, implicitHook)
+	}
+
+	return allHooks
+}
+
+func (h *CommandHooks) getExplicitHooks(prefix HookType, validHookNames []string) []*ScriptConfig {
+	matchingScripts := []*ScriptConfig{}
+
+	// Find explicitly configured hooks from azure.yaml
+	for scriptName, scriptConfig := range h.scripts {
+		if scriptConfig == nil {
+			continue
+		}
+
+		index := slices.IndexFunc(validHookNames, func(hookName string) bool {
+			return hookName == scriptName
+		})
+
+		if index > -1 {
+			// If the script config includes an OS specific configuration use that instead
+			if runtime.GOOS == "windows" && scriptConfig.Windows != nil {
+				scriptConfig = scriptConfig.Windows
+			} else if (runtime.GOOS == "linux" || runtime.GOOS == "darwin") && scriptConfig.Linux != nil {
+				scriptConfig = scriptConfig.Linux
+			}
+
+			scriptConfig.Name = scriptName
+			scriptConfig.Path = strings.ReplaceAll(scriptConfig.Path, "/", string(os.PathSeparator))
+			matchingScripts = append(matchingScripts, scriptConfig)
+		}
+	}
+
+	return matchingScripts
+}
+
+func (h *CommandHooks) getImplicitHooks(prefix HookType, validHookNames []string) []*ScriptConfig {
+	matchingScripts := []*ScriptConfig{}
+
+	hooksDir := filepath.Join(h.cwd, ".azure", "hooks")
```

Comment:
Nothing that if we do proceed with implicit hooks, we need to modify `.gitignore` in our templates and examples to not simply ignore everything under `.gitignore`.

Reply:
Agreed - we may want to reduce the .gitignore to allow files in the `.azure` folder or choose a different default location.

Reply:
Users can also manually force adds that would typically be ignored by .gitignore.